### ISSUE TITLE
[Feature] Group dependent hyperparameters for sweep efficiency

### ISSUE THREAD

**Issue by ameyagodbole** [wandb/wandb#2853]
**Is your feature request related to a problem? Please describe.**
It is possible to have certain hyperparameters have no effect based on flags set by other hyperparameters (especially categorical)
e.g. if I have a categorical hyperparam `loss_fn` choosing between margin loss and cross-entropy
```yaml
loss_fn:
  values:
  - "margin"
  - "cross-entropy"
```
 and (float) hyperparams `margin` and `temperature` for the two losses respectively
```yaml
margin:
  min: 0.0
  max: 10.0

temperature:
  distribution: log_uniform
  min: -7.0
  max: 0.0
```
then based on the `loss_fn`, one param out of `margin` and `temperature` will be unused.

As far as I know, there is no way to specify these parameter groups. I think having such a nested hyperparameter definition would allow the tuning algorithms to be more efficient:
1. Grid-search/Random-search can avoid redundant configs such as:
`--loss_fn=margin --margin=0.5 --temperature=0.1` and `--loss_fn=margin --margin=0.5 --temperature=1.0`
2. Bayes gets information about all possible values of `temperature` for a given setting of `--loss_fn=margin --margin=0.5`

Is this something W&B could potentially have?

**Describe the solution you'd like**
Being able to define a hierarchy of hyperparameters such that a higher-level hyperparameter marks out certain lower-level hyperparameters as redundant.
E.g. (This is a suggested format given my very limited familiarity with YAML)
```yaml
loss_fn:
  values:
  - "margin"
  - "cross-entropy"

margin:
  min: 0.0
  max: 10.0
  requires:
    param: loss_fn
    value: "margin"

temperature:
  distribution: log_uniform
  min: -7.0
  max: 0.0
  requires:
    param: loss_fn
    value: "cross-entropy"
```

The sweep controller can then sample configs iteratively starting from the least constrained parameters and avoid redundant configs. For Bayes, it can populate additional samples for unused hyperparameters.

**Describe alternatives you've considered**
For running the sweep, there's no real solution as such; either run the sweep long enough or create a different sweep for every combination of higher-level categorical hyperparameters. But if you have multiple groups such as this, it gets out of hand quickly. For Bayesian search, it also prevents the reuse of samples across sweeps.

For analysis, it is always possible to see parameter importance and results after filtering based on combinations of higher-level categorical hyperparameters. 

**Additional context**
-

**Comment by anmolmann** [comment]
Hey @ameyagodbole, looking into this for you.

**Comment by anmolmann** [comment]
@ameyagodbole, thanks for providing details for your use-case. I think what you might be looking for is something on the lines of conditional sampling/param choices. We do have a feature request filed for this and I’ve bumped up the priority for it too. We’ll keep you posted once there’s an update.

Meanwhile, one way to address this problem for now would be to pre-process the sweep.yaml (or the sweep config) and provide plausible combinations of pairs of (`param1`, `param2`) that would satisfy your condition (which you can do because of your prior knowledge re: the use-case).

**Comment by ameyagodbole** [comment]
Glad to know that this feature is already requested. Yes, this would not only be conditional sampling, but for the case of Bayesian tuning, it would provide additional samples (at no compute cost) before deciding the next config to try.

I did not know about the paired hyperparameters. Is there documentation for this that I can refer to?

**Comment by ibash** [comment]
For posterity, I think @anmolmann means making a script like this:

sweep.yml
```yaml
program: train.py
method: grid
metric:
  goal: minimize
  name: val_loss
parameters:
  num_convolutions:
    min: 1
    max: 4

  num_dense:
    min: 1
    max: 4

  # these get autogenerated in make_sweeps.py
  filters_n:
    min: 1
    max: 5

  kernel_size_n:
    min: 1
    max: 5

  pool_size_n:
    min: 1
    max: 5

  dense_n:
    min: 1
    max: 5
```

make_sweeps.py
```python
import copy
import itertools
import pathlib
import yaml

path = pathlib.Path('./sweep.yml')
with path.open() as file:
    sweep = yaml.safe_load(file)
    parameters = sweep['parameters']

    num_convolutions = parameters['num_convolutions']
    range1 = range(num_convolutions['min'], num_convolutions['max'] + 1)


    num_dense = parameters['num_dense']
    range2 = range(num_dense['min'], num_dense['max'] + 1)

    combinations = itertools.product(range1, range2)

    for (num_convolutions, num_dense) in combinations:
        instance = copy.deepcopy(sweep)
        parameters = instance['parameters']

        del parameters['num_convolutions']['min']
        del parameters['num_convolutions']['max']
        parameters['num_convolutions']['value'] = num_convolutions

        for key in ['filters', 'kernel_size', 'pool_size']:
            for i in range(num_convolutions):
                parameters[f'{key}_{i+2}'] = copy.deepcopy(parameters[f'{key}_n'])
            del parameters[f'{key}_n']

        del parameters['num_dense']['min']
        del parameters['num_dense']['max']
        parameters['num_dense']['value'] = num_dense

        for i in range(num_dense):
            parameters[f'dense_{i+1}'] = copy.deepcopy(parameters['dense_n'])
        del parameters['dense_n']
       
        print(yaml.dump(instance, sort_keys=False))

```

**Comment by guillem-ms** [comment]
Hi! Any news on this feature?

**Comment by anmolmann** [comment]
Hi @guillem-ms! Unfortunately, this is still in our backlog. I'll raise this internally again and follow up with you soon!

**Comment by guillem-ms** [comment]
Thanks @anmolmann, that would be great :) 
It would be very interesting for many workflows.

**Comment by luisbergua** [comment]
Hey all, I just wanted to let you know that conditional sweeps is possible using Launch, check out [this](https://colab.research.google.com/drive/1WxLKaJlltThgZyhc7dcZhDQ6cjVQDfil#scrollTo=AFEzIxA6foC7) colab.
